subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java

```diff
@@ -146,6 +151,25 @@ public BuildOperationTrace(StartParameter startParameter, BuildOperationListener
         buildOperationListenerManager.addListener(listener);
     }
 
+    private static String getProperty(StartParameter startParameter, String property) {
+        Map<String, String> sysProps = startParameter.getSystemPropertiesArgs();
+        String basePath = sysProps.get(property);
+        if (basePath == null) {
+            basePath = System.getProperty(property);
+        }
+        return basePath;
+    }
+
+    @Nullable
+    private static Set<String> getFilter(StartParameter startParameter) {
```

Comment:
❓ Why not return an empty `Set` here instead of a potential `null`?  I think it's rare to need to return `null` from a method returning a collection.

This would allow many null checks to be removed from `LoggingBuildOperationListener`, too.  

If not removed, I think more `@Nullable` annotations are needed along the data flow of where this result is used.

Reply:
Added `FilteringBuildOperationListener` to avoid null checks